libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/idp/IDPAccountPickerActivity.java

```diff
@@ -0,0 +1,161 @@
+
+/**
+ * This class provides UI to select an existing signed in user account, or add a new account.
+ * It kicks off the IDP authentication flow once the selection is made. This screen is popped
+ * off the activity stack once the flow is complete, passing back control to the SP app.
+ *
+ * @author bhariharan
+ */
+public class IDPAccountPickerActivity extends AccountSwitcherActivity {
+
+    private static final int IDP_LOGIN_REQUEST_CODE = 999;
+    private static final String TAG = "IDPAccountPickerActivity";
+
+    private SPConfig spConfig;
+
+    @Override
+    public void onCreate(Bundle savedInstance) {
+        super.onCreate(savedInstance);
+        final Button button = findViewById(R.id.sf__switcher_apply_button);
+        button.setText(R.string.sf__idp_account_selector);
+
+        // Fetches the required extras.
+        final Intent intent = getIntent();
+        final Bundle extras = intent.getExtras();
+        if (extras != null) {
+            spConfig = new SPConfig(extras.getBundle(IDPCodeGeneratorActivity.SP_CONFIG_BUNDLE_KEY));
+        }
+    }
+
+    @Override
+    protected List<UserAccount> getAccounts() {
+        final List<UserAccount> accounts = userAccMgr.getAuthenticatedUsers();
+
+        // If no login server is passed in, return all user accounts.
+        if (TextUtils.isEmpty(spConfig.getLoginUrl())) {
+            return accounts;
+        }
+        final List<UserAccount> filteredAccounts = new ArrayList<>();
+        if (accounts != null) {
+            for (final UserAccount account : accounts) {
+
+                // If user account has the same login server, add it to the list.
+                if (spConfig.getLoginUrl().equals(account.getLoginServer())) {
+                    filteredAccounts.add(account);
+                }
+            }
+        }
+        if (filteredAccounts.size() == 0) {
+            return null;
+        }
+        return filteredAccounts;
+    }
+
+    @Override
+    protected void accountSelected(UserAccount account) {
+        SalesforceSDKLogger.d(TAG, "Account selected: " + account);
+
+        /*
+         * If an account is passed in, proceeds with IDP auth flow directly. This will kick off
+         * authentication for the SP app without messing with the state of the IDP. If no account
+         * is passed in, kicks off new user login flow, sets current user on the IDP to the newly
+         * logged in user, and then resumes authentication for the SP app with that user.
+         */
+        if (account != null) {
+            proceedWithIDPAuthFlow(account);
+        } else {
+            kickOffNewUserFlow();
+        }
+    }
+
+    @Override
+    protected void finishActivity() {
+
+        // Do nothing in here, since we don't want to finish this activity until the IDP flow is complete.
+    }
+
+    @Override
+    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
+        if (requestCode == IDP_LOGIN_REQUEST_CODE) {
```

Comment:
Handles results based on what was triggered. If login for IDP was completed, trigger IDP code flow with that user. If IDP code flow was completed, pass on the results to the SP app and finish this activity.